File: sdk/elixir/lib/rocksky/playlist.ex
defmodule Rocksky.Playlist do
  @moduledoc """
  `app.rocksky.playlist.*` — the global, AT-Proto-backed playlists.

  Distinct from `Rocksky.Library`, whose playlist functions drive the
  Subsonic/Navidrome library (`app.rocksky.library.*`).

  Writes publish records to the caller's repo, so they take a `token` and only
  appear in reads once the AppView has ingested the commit.
  """

  defp to_bin(nil), do: ""
  defp to_bin(v) when is_binary(v), do: v
  defp to_bin(v), do: to_string(v)

  @doc "The playlist catalog."
  def list(limit \\ 50, offset \\ 0, base \\ ""),
    do: :rocksky.playlists(to_bin(base), limit, offset)

  @doc "A single playlist with its tracks."
  def get(uri, base \\ ""), do: :rocksky.playlist(to_bin(base), to_bin(uri))

  @doc """
  Create a playlist. Returns `%{"uri" => ..., "cid" => ...}`. Pass `""` to omit
  an optional field.
  """
  def create(token, name, description \\ "", picture_url \\ "", base \\ ""),
    do:
      :rocksky.create_playlist(
        to_bin(base),
        to_bin(token),
        to_bin(name),
        to_bin(description),
        to_bin(picture_url)
      )

  @doc "Rename or re-describe a playlist. Owner only; the AT-URI is unchanged."
  def update(token, uri, name \\ "", description \\ "", picture_url \\ "", base \\ ""),
    do:
      :rocksky.update_playlist(
        to_bin(base),
        to_bin(token),
        to_bin(uri),
        to_bin(name),
        to_bin(description),
        to_bin(picture_url)
      )

  @doc """
  Add songs by their `app.rocksky.song` AT-URIs. Owner only. Returns the AT-URIs
  of the created entry records.
  """
  def add_songs(token, uri, songs, base \\ ""),
    do:
      :rocksky.add_songs_to_playlist(
        to_bin(base),
        to_bin(token),
        to_bin(uri),
        Enum.map(songs, &to_bin/1)
      )

  @doc "Remove a song. Only the repo that added an entry can retract it."
  def remove_track(token, uri, song_uri, base \\ ""),
    do:
      :rocksky.remove_playlist_track(
        to_bin(base),
        to_bin(token),
        to_bin(uri),
        to_bin(song_uri)
      )

  @doc "Delete a playlist and the caller's own entries. Owner only."
  def remove(token, uri, base \\ ""),
    do: :rocksky.remove_playlist(to_bin(base), to_bin(token), to_bin(uri))

  @doc """
  Escape hatch — call any `app.rocksky.*` procedure whose arguments ride the
  query string.
  """
  def post(nsid, params \\ %{}, base \\ "", token \\ ""),
    do: :rocksky.post(to_bin(nsid), params, to_bin(base), to_bin(token))
end

File: sdk/rust/src/resources/playlist.rs
//! `app.rocksky.playlist.*` — playlist CRUD.

use serde::Serialize;
use serde_json::Value;

use crate::client::Client;
use crate::error::Result;
use crate::models::{Playlist, PlaylistBasic, PlaylistsEnvelope};

/// The AT-URI + CID of a playlist record returned by create/update.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct PlaylistRef {
    pub uri: String,
    pub cid: String,
}

/// AT-URIs of the `app.rocksky.playlist.song` entries created by `add_songs`.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct AddedSongs {
    pub uris: Vec<String>,
}

#[derive(Debug)]
pub struct PlaylistApi<'a> {
    client: &'a Client,
}

impl<'a> PlaylistApi<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Self { client }
    }

    pub async fn get(&self, uri: impl Into<String>) -> Result<Playlist> {
        #[derive(Serialize)]
        struct P {
            uri: String,
        }
        self.client
            .query_as(
                "app.rocksky.playlist.getPlaylist",
                &P { uri: uri.into() },
                false,
            )
            .await
    }

    pub fn list(&self) -> ListPlaylists<'_> {
        ListPlaylists {
            client: self.client,
            params: Page {
                limit: None,
                offset: None,
            },
        }
    }

    /// Publishes an `app.rocksky.playlist` record to the caller's repo. The
    /// AppView only lists the playlist once the commit has been ingested, so a
    /// read straight after this may not see it yet.
    pub async fn create(
        &self,
        name: impl Into<String>,
        description: Option<String>,
        picture_url: Option<String>,
    ) -> Result<PlaylistRef> {
        #[derive(Serialize)]
        #[serde(rename_all = "camelCase")]
        struct P {
            name: String,
            #[serde(skip_serializing_if = "Option::is_none")]
            description: Option<String>,
            #[serde(skip_serializing_if = "Option::is_none")]
            picture_url: Option<String>,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.createPlaylist",
                Some(&P {
                    name: name.into(),
                    description,
                    picture_url,
                }),
                None::<&()>,
                true,
            )
            .await
    }

    /// Rename or re-describe a playlist. Owner only — the record is rewritten
    /// on its existing rkey, so the AT-URI is stable.
    pub async fn update(
        &self,
        uri: impl Into<String>,
        name: Option<String>,
        description: Option<String>,
        picture_url: Option<String>,
    ) -> Result<PlaylistRef> {
        #[derive(Serialize)]
        #[serde(rename_all = "camelCase")]
        struct P {
            uri: String,
            #[serde(skip_serializing_if = "Option::is_none")]
            name: Option<String>,
            #[serde(skip_serializing_if = "Option::is_none")]
            description: Option<String>,
            #[serde(skip_serializing_if = "Option::is_none")]
            picture_url: Option<String>,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.updatePlaylist",
                Some(&P {
                    uri: uri.into(),
                    name,
                    description,
                    picture_url,
                }),
                None::<&()>,
                true,
            )
            .await
    }

    /// Add songs by their `app.rocksky.song` AT-URIs. Owner only. Returns the
    /// AT-URIs of the created `app.rocksky.playlist.song` entries.
    pub async fn add_songs(
        &self,
        uri: impl Into<String>,
        songs: Vec<String>,
    ) -> Result<AddedSongs> {
        #[derive(Serialize)]
        struct P {
            uri: String,
            songs: Vec<String>,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.addSongs",
                Some(&P {
                    uri: uri.into(),
                    songs,
                }),
                None::<&()>,
                true,
            )
            .await
    }

    /// Remove a song from a playlist. An entry record lives in the repo that
    /// published it, so only that repo can retract it.
    pub async fn remove_track(
        &self,
        uri: impl Into<String>,
        song_uri: impl Into<String>,
    ) -> Result<Value> {
        #[derive(Serialize)]
        #[serde(rename_all = "camelCase")]
        struct P {
            uri: String,
            song_uri: String,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.removeTrack",
                Some(&P {
                    uri: uri.into(),
                    song_uri: song_uri.into(),
                }),
                None::<&()>,
                true,
            )
            .await
    }

    pub async fn remove(&self, uri: impl Into<String>) -> Result<Value> {
        #[derive(Serialize)]
        struct P {
            uri: String,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.removePlaylist",
                Some(&P { uri: uri.into() }),
                None::<&()>,
                true,
            )
            .await
    }

    pub async fn start(
        &self,
        uri: impl Into<String>,
        shuffle: Option<bool>,
        position: Option<u32>,
    ) -> Result<Value> {
        #[derive(Serialize)]
        #[serde(rename_all = "camelCase")]
        struct P {
            uri: String,
            #[serde(skip_serializing_if = "Option::is_none")]
            shuffle: Option<bool>,
            #[serde(skip_serializing_if = "Option::is_none")]
            position: Option<u32>,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.startPlaylist",
                Some(&P {
                    uri: uri.into(),
                    shuffle,
                    position,
                }),
                None::<&()>,
                true,
            )
            .await
    }

    pub async fn insert_files(
        &self,
        uri: impl Into<String>,
        files: Vec<String>,
        position: Option<u32>,
    ) -> Result<Value> {
        #[derive(Serialize)]
        struct P {
            uri: String,
            files: Vec<String>,
            #[serde(skip_serializing_if = "Option::is_none")]
            position: Option<u32>,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.insertFiles",
                Some(&P {
                    uri: uri.into(),
                    files,
                    position,
                }),
                None::<&()>,
                true,
            )
            .await
    }

    pub async fn insert_directory(
        &self,
        uri: impl Into<String>,
        directory: impl Into<String>,
        position: Option<u32>,
    ) -> Result<Value> {
        #[derive(Serialize)]
        struct P {
            uri: String,
            directory: String,
            #[serde(skip_serializing_if = "Option::is_none")]
            position: Option<u32>,
        }
        self.client
            .procedure_as(
                "app.rocksky.playlist.insertDirectory",
                Some(&P {
                    uri: uri.into(),
                    directory: directory.into(),
                    position,
                }),
                None::<&()>,
                true,
            )
            .await
    }
}

#[derive(Debug, Serialize)]
struct Page {
    #[serde(skip_serializing_if = "Option::is_none")]
    limit: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    offset: Option<u32>,
}

#[derive(Debug)]
pub struct ListPlaylists<'a> {
    client: &'a Client,
    params: Page,
}

impl<'a> ListPlaylists<'a> {
    pub fn limit(mut self, limit: u32) -> Self {
        self.params.limit = Some(limit);
        self
    }
    pub fn offset(mut self, offset: u32) -> Self {
        self.params.offset = Some(offset);
        self
    }
    pub async fn send(self) -> Result<Vec<PlaylistBasic>> {
        let env: PlaylistsEnvelope = self
            .client
            .query_as("app.rocksky.playlist.getPlaylists", &self.params, false)
            .await?;
        Ok(env.playlists)
    }
}
